packages/typespec-vscode/src/vscode-cmd/emit-code/emit-code.ts

```diff
@@ -153,60 +149,82 @@ async function doEmit(mainTspFile: string, emitter: Emitter) {
+            : "";
+          let packageFullName = emitter.package;
+          if (version) {
+            packageFullName = `${emitter.package}@${version}`;
+          }
+          packageQuickPickItems.push({
+            label: `${emitter.package}`,
+            description: `TypeSpec library for emitting ${emitter.language} from TypeSpec files.`,
+            detail: minimumRequisites,
+            packageFullName: packageFullName,
+            sourceRepo: emitter.sourceRepo,
+            picked: true,
+            buttons: emitter.sourceRepo
+              ? [
+                  {
+                    iconPath: new vscode.ThemeIcon("link-external"),
+                    tooltip: "More details",
+                    uri: emitter.sourceRepo,
+                  },
+                ]
+              : undefined,
+          });
+          packagesToInstall.push({ name: emitter.package, version: version });
         }
-        packageQuickPickItems.push({
-          label: `${emitter.package}`,
-          description: `TypeSpec library for emitting ${emitter.language} from TypeSpec files.`,
-          detail: minimumRequisites,
-          packageFullName: packageFullName,
-          sourceRepo: emitter.sourceRepo,
-          picked: true,
-          buttons: emitter.sourceRepo
-            ? [
-                {
-                  iconPath: new vscode.ThemeIcon("link-external"),
-                  tooltip: "More details",
-                  uri: emitter.sourceRepo,
-                },
-              ]
-            : undefined,
-        });
-        packagesToInstall.push(packageFullName);
-      }
-
-      for (const p of packagesToInstall) {
-        /* verify dependency packages. */
-        try {
-          const dependenciesToInstall = await npmUtil.calculateNpmPackageDependencyToUpgrade(
-            p,
-            version,
-            [npmDependencyType.dependencies, npmDependencyType.peerDependencies],
-          );
-          if (dependenciesToInstall.length > 0) {
-            packagesToInstall.push(...dependenciesToInstall.map((d) => `${d.name}@${d.version}`));
-            for (const dep of dependenciesToInstall) {
-              const packageFullName = `${dep.name}@${version ?? "latest"}`;
-              packageQuickPickItems.push({
-                label: `${dep.name}`,
-                description: "Required for enabling the emitter library.",
-                packageFullName: packageFullName,
-                picked: true,
-              });
+        for (const p of packagesToInstall) {
+          /* verify dependency packages. */
+          try {
+            const dependenciesToInstall = (
+              await npmUtil.calculateNpmPackageDependencyToUpgrade(p.name, p.version, [
+                npmDependencyType.dependencies,
+                npmDependencyType.peerDependencies,
+              ])
+            ).filter((d) => !packagesToInstall.includes(d));
```

Comment:
what if they have the same dependency but not in packagesToInstall array?

Reply:
packagesToInstall track all the packages needed to install/upgrade. This line code is to filter out the already existed package. So we will check all the dependency package